crawl.sh

```diff
@@ -1,16 +1,73 @@
 #!/usr/bin/env bash
+
+# Recursively crawl a website and save its HTML locally.
+#
+# Example usage:
+#
+# ./crawl.sh [-d depth] https://www.consumerfinance.gov
+#
+# Optionally specify -d depth to limit the crawl depth.
+
+# Fail if anything fails.
+set -e
+
+depth=0
+
+while getopts ":d:" opt; do
+    case $opt in
+      d )
+        depth="$OPTARG";
+        number_regex='^[0-9]+$'
+        if ! [[ $depth =~ $number_regex ]] ; then
+            echo "Crawl depth must be a number." 1>&2
+            exit 1
+        fi
+        ;;
+    \? )
+        echo "Invalid option: -$OPTARG." 1>&2
+        exit 1
+        ;;
+    : )
+        echo "Invalid option: -$OPTARG requires an argument." 1>&2
+        exit 1
+        ;;
+    esac
+done
+
+shift $((OPTIND -1))
+
+url=$1
+
+if [ -z "$url" ]; then
+  echo "Must specify URL to crawl."
+  exit 1
+fi
+
+echo "Starting crawl at $url."
+
+if [ $depth -ne 0 ]; then
+    echo "Limiting crawl to depth $depth."
+fi
+
+domain=$url
+domain="${domain#http://}"
+domain="${domain#https://}"
+domain="${domain%%:*}"
+domain="${domain%%\?*}"
+domain="${domain%%/*}"
+echo "Limiting crawl to domain $domain."
+
 time wget \
-    --domains=www.consumerfinance.gov \
-    --exclude-domains=files.consumerfinance.gov \
+    --domains="$domain" \
     --execute robots=off \
     --follow-tags=a \
-    --limit-rate=200k \
-    --random-wait \
+    --limit-rate=1m \
     --accept html \
     --reject-regex "topics=|authors=|categories=|filter_blog_category=|ext_url=|search_field=|issuer_name=" \
     --recursive \
-    --level=4 \
+    --level="$depth" \
     --trust-server-names \
-    --no-verbose \
+    --verbose \
     --no-clobber \
-    https://www.consumerfinance.gov/
+    --rejected-log=rejected.log \
```

Comment:
Nice improvements to the logging.

Reply:
Unfortunately, this rejection log can be quite large! On [my most recent run](https://github.com/chosak/crawl-cfgov/runs/1333362105?check_suite_focus=true) it caused the commit back to fail:

```
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.        
remote: error: Trace: 7c30218a339d6765915958ab1655bf7f9f0704cb19e9a2505fba2079455c2f07        
remote: error: See http://git.io/iEPt8g for more information.        
remote: error: File rejected.log is 147.20 MB; this exceeds GitHub's file size limit of 100.00 MB        
To https://github.com/chosak/crawl-cfgov
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to 'https://github.com/chosak/crawl-cfgov'
```

I think having this information would be very useful though. Before trying Git LFS, I'll try gzipping the logfiles before committing them back.

Reply:
This worked on my fork, added here in 980aff8.